cli/flox/src/commands/check_for_upgrades.rs

```diff
@@ -0,0 +1,298 @@
+use std::fs::{self, File};
+use std::path::{Path, PathBuf};
+use std::process::{Command, Stdio};
+use std::time::SystemTime;
+
+use anyhow::{bail, Context, Result};
+use bpaf::{Bpaf, Parser};
+use flox_rust_sdk::flox::Flox;
+use flox_rust_sdk::models::environment::Environment;
+use flox_rust_sdk::providers::upgrade_checks::{UpgradeInformation, UpgradeInformationGuard};
+use flox_rust_sdk::utils::CommandExt;
+use serde::de::DeserializeOwned;
+use tracing::{debug, info_span, instrument};
+
+use super::UninitializedEnvironment;
+use crate::subcommand_metric;
+
+/// By default check once a day
+const DEFAULT_TIMEOUT_SECONDS: u64 = 24 * 60 * 60;
+
+#[derive(Bpaf, Clone)]
+pub struct CheckForUpgrades {
+    /// Skip checking for upgrade if checked less <timeout> seconds ago
+    #[bpaf(long, argument("seconds"), fallback(DEFAULT_TIMEOUT_SECONDS))]
+    check_timeout: u64,
+
+    #[bpaf(external(parse_uninitialized_environment_json))]
+    environment: UninitializedEnvironment,
+}
+
+fn parse_uninitialized_environment_json<T: DeserializeOwned>() -> impl Parser<T> {
+    bpaf::positional("environment")
+        .help("JSON representation of an uninitialized environment to be checked")
+        .parse(|string: String| serde_json::from_str::<T>(&string))
+}
+
+#[derive(Debug, PartialEq)]
+enum ExitBranch {
+    LockTaken,
+    AlreadyChecked,
+    Checked,
+}
+
+impl CheckForUpgrades {
+    #[instrument(name = "check-upgrade", skip_all)]
+    pub async fn handle(self, flox: Flox) -> Result<()> {
+        subcommand_metric!("check-upgrade");
+        self.check_for_upgrades(&flox)?;
+        Ok(())
+    }
+
+    fn check_for_upgrades(self, flox: &Flox) -> Result<ExitBranch> {
+        let mut environment = self.environment.into_concrete_environment(flox)?;
+
+        let upgrade_information = UpgradeInformationGuard::read_in(environment.cache_path()?)?;
+
+        // Return if previous information
+        // - exists &&
+        // - targets the current lockfile &&
+        // - has recently been fetched
+        // Otherwise, run a dry-upgrade tof he environment and store the new information
+        if let Some(info) = upgrade_information.info() {
+            let environment_lockfile = environment.lockfile(flox)?;
+
+            let is_information_for_current_lockfile =
+                info.result.old_lockfile == Some(environment_lockfile);
+            let is_checked_recently =
+                info.last_checked.elapsed().unwrap().as_secs() <= self.check_timeout;
```

Comment:
nit: probably doesn't matter but if I set `check_timeout` to 0 I would expect the check to re-run, but I think this would have to be `<` instead of `<=` for that to be the case

Reply:
yes `<` should be alright, i think i had some trouble with tests earlier on that made me set `<=` but that isnt the case anymore.
Changing to `<` then.